src/main/java/io/nats/client/impl/MessageQueue.java

```diff
@@ -344,37 +382,43 @@ long sizeInBytes() {
         return sizeInBytes.get();
     }
 
-    void filter(Predicate<NatsMessage> p) {
-        editLock.lock();
+    void filterOnStop() {
+        if (this.isRunning()) {
+            throw new IllegalStateException("Filter is only supported when the queue is paused");
+        }
+        countLock.lock();
         try {
-            if (this.isRunning()) {
-                throw new IllegalStateException("Filter is only supported when the queue is paused");
-            }
-            ArrayList<NatsMessage> newQueue = new ArrayList<>();
-            NatsMessage cursor = this.queue.poll();
-            while (cursor != null) {
-                if (!p.test(cursor)) {
-                    newQueue.add(cursor);
-                } else {
-                    sizeInBytes.addAndGet(-cursor.getSizeInBytes());
-                    length.decrementAndGet();
+            ArrayList<NatsMessage> tempQueue = new ArrayList<>();
+            this.queue.drainTo(tempQueue); // I realize this is a copy of a queue, but polling locks the queue every time
+            long removed = 0;
+            long bytesRemoved = 0;
+            for (NatsMessage msg : tempQueue) {
+                if (msg instanceof ProtocolMessage) {

```

Comment:
All ProtocolMessages are filtered on stop, they are the only type of messages that are.